Add Azure IoT Edge integration (#7465)
DataDog/integrations-core: azure_iot_edge/README.md

# Agent Check: Azure IoT Edge

## Overview

> [Azure IoT Edge][1] is a fully managed service to deploy Cloud workloads to run on Internet of Things (IoT) edge devices via standard containers.

Reviewer: ```suggestion
[Azure IoT Edge][1] is a fully managed service to deploy Cloud workloads to run on Internet of Things (IoT) Edge devices via standard containers.
```

Use the Datadog-Azure IoT Edge integration to collect metrics and health status from IoT Edge devices.

> **Note**: this integration requires IoT Edge runtime version 1.0.10 or above.

Reviewer: ```suggestion
**Note**: This integration requires IoT Edge runtime version 1.0.10 or above.
```

## Setup

Follow the instructions below to install and configure this check for an IoT Edge device running on a device host.

### Installation

The Azure IoT Edge check is included in the [Datadog Agent][2] package.

No additional installation is needed on your device.

### Configuration

> It is recommended to configure the IoT Edge device so that the Agent runs as a custom module. Follow the official Microsoft documentation on [deploying Azure IoT Edge modules][3] for general information on installing and working with custom modules for Azure IoT Edge.

Reviewer: ```suggestion
Configure the IoT Edge device so that the Agent runs as a custom module. Follow the Microsoft documentation on [deploying Azure IoT Edge modules][3] for information on installing and working with custom modules for Azure IoT Edge.
```

Follow the steps below to configure the IoT Edge device, runtime modules, and the Datadog Agent to start collecting IoT Edge metrics.

1. Configure the **Edge Agent** runtime module as follows:
    - Image version must be `1.0.10` or above.
    - Under "Create Options", add the following `Labels`. Edit the `com.datadoghq.ad.instances` label as appropriate. See the [sample azure_iot_edge.d/conf.yaml][5] for all available configuration options. See the documentation on [Docker Integrations Autodiscovery][6] for more information on labels-based integration configuration.

        ```json
        "Labels": {
            "com.datadoghq.ad.check_names": "[\"azure_iot_edge\"]",
            "com.datadoghq.ad.init_configs": "[{}]",
            "com.datadoghq.ad.instances": "[{\"edge_hub_prometheus_url\": \"http://edgeHub:9600/metrics\", \"edge_agent_prometheus_url\": \"http://edgeAgent:9600/metrics\"}]"
        }
        ```

>     - Under "Environment Variables", experimental metrics must be enabled by adding these environment variables (note the double underscores):

Reviewer: ```suggestion
    - Under "Environment Variables", enable experimental metrics by adding these environment variables (note the double underscores):
```

        - `ExperimentalFeatures__Enabled`: `true`
        - `ExperimentalFeatures__EnableMetrics`: `true`

1. Configure the **Edge Hub** runtime module as follows:
    - Image version must be `1.0.10` or above.

>     - Under "Environment Variables", experimental metrics must be enabled by adding these environment variables (note the double underscores):

Reviewer: ```suggestion
    - Under "Environment Variables", enable experimental metrics by adding these environment variables (note the double underscores):
```

Use the Datadog-Azure IoT Edge integration to collect metrics and health status from IoT Edge devices.

## Setup

Follow the instructions below to install and configure this check for an IoT Edge device running on a device host.

### Installation

The Azure IoT Edge check is included in the [Datadog Agent][2] package.

No additional installation is needed on your device.

### Configuration

It is recommended to configure the IoT Edge device so that the Agent runs as a custom module. Follow the official Microsoft documentation on [deploying Azure IoT Edge modules][3] for general information on installing and working with custom modules for Azure IoT Edge.

Follow the steps below to configure the IoT Edge device, runtime modules, and the Datadog Agent to start collecting IoT Edge metrics.

1. Edit your IoT Edge device `config.yaml` file:
    - **Linux**: Make sure the `connect.management_uri` and `listen.management_uri` options point to a Unix Domain Socket (note that this is the default configuration). For example:

        ```yaml
        # /etc/iotedge/config.yaml

        connect:
            management_uri: "unix:///var/run/iotedge/mgmt.sock"
            # ...

        listen:

>             management_uri: "fd://iotedge.mgmt.socket"

Author: Yes. The default is `fd://...` which actually corresponds to the `systemd` format. (The Linux Security Manager is actually just a systemd service, and it does its own socket binding thing, this is all related to how the Security Manager is configured by default. I can dive into specifics if needed).

I preferred to keep the default config rather than requiring users to set eg `/var/run/iotedge/mgmt.sock` explicitly, although I guess that's a possible option (but that would mean an extra thing for users to tweak). The config listed in the Linux case here is 100% the default one, untouched, which is why I wrote "Make sure that…".

>             management_uri: "fd://iotedge.mgmt.socket"

Author: Now irrelevant since this config was dropped, as we dropped the security manager service check.

            # ...
        ```

>     - **Windows**: Make sure the `connect.management_uri` and `listen.management_uri` options point to an HTTP endpoint. For example:

Author: No, on Windows that's not the default configuration. And I'm actually having issues with setting this up on Windows (whereas it does work quite fine on Linux - see our test setup).

The reason behind this diff is that we can't support connecting through a Unix Socket on Windows for now (due to a Python limitation: Python just hasn't implemented `AF_UNIX` on Windows yet). So we need a different strategy to get the status of the Security Manager on Windows.

Until now I thought that connecting through a regular TCP socket would do the trick. Unfortunately my testing yesterday shows that this is not trivial, as in the default installation IoT Edge ships with a stripped down version of Docker where eg `host.docker.internal` is not available, so I'm having issues setting `connect.management_uri` in a way that allows the Edge Agent container to talk to the Security Manager on the host. 🤷 

Other options I'm considering are:

- Drop the `security_manager.can_connect` service check - either on Windows only, or completely. Rationale: if the Edge Agent and Edge Hub are up and running, then that certainly means the Security Manager is fine too (since it's the one managing those two components).
- Find another way to check the Security Manager health on Windows, knowing that the Datadog Agent would have to reach to the host from a container. (But I'm not sure what that would be, any ideas? The Security Manager is a Windows service.)

        ```yaml

>         # /etc/iotedge/config.yaml

Author: No, good catch :+1:

>         # /etc/iotedge/config.yaml

Author: Now irrelevant.

    - Image version must be `1.0.10` or above.
    - Under "Environment Variables", experimental metrics must be enabled by adding these environment variables (note the double underscores):
        - `ExperimentalFeatures__Enabled`: `true`
        - `ExperimentalFeatures__EnableMetrics`: `true`

1. Install and configure the Datadog Agent as a **custom module**:
    - Set the module name. For example: `datadog-agent`.
    - Set the Agent image URI. For example: `datadog/agent:7.24.0`.
    - Under "Environment Variables", configure your `DD_API_KEY`. You may also set extra Agent configuration here (see [Agent Environment Variables][4]).
    - Under "Container Create Options", enter the following configuration based on your device OS. Edit the `com.datadoghq.ad.instances` option as appropriate. See the [sample azure_iot_edge.d/conf.yaml][5] for all available configuration options. See the documentation on [Docker Integrations Autodiscovery][6] for more information on labels-based integration configuration. **Note**: `NetworkId` must correspond to the network name set in the device `config.yaml` file.
        - Linux:
            ```json
            {
                "HostConfig": {

>                     "NetworkMode": "default",

Author: What do you mean by explicit?

> The agent runs in bridge mode by default I believ

Perhaps, but on Windows Containers there is no `bridge` network driver. Eg in the installation I've done yesterday, the network mode is NAT. See [Windows containers network drivers](https://docs.microsoft.com/en-us/virtualization/windowscontainers/container-networking/network-drivers-topologies)

To be honest I am not sure about this `default` network mode value, and need to check what it does exactly. That said, I was able to get the integration working on both Linux VM and Windows VM with the settings documented in the README currently (except for the mgmt API problem on Windows).

>                     "NetworkMode": "default",

Reviewer 2: That's fine if it's the default. I meant specifying "bridge" or "host" for example. But let's keep the config as close as the default one 👍

                    "Env": ["NetworkId=azure-iot-edge"],
                    "Binds": [
                        "/var/run/docker.sock:/var/run/docker.sock",
                        "/var/run/iotedge/mgmt.sock:/var/run/iotedge/mgmt.sock"
                    ]
                },
                "Labels": {

>                     "com.datadoghq.ad.check_names": "[\"azure_iot_edge\"]",

Author: Right. I went off from https://docs.datadoghq.com/agent/docker/integrations/ and applied the approach there to the Datadog Agent container itself, but I'd agree that in retrospect it's a bit "exotic".

I could add these labels to an arbitrary container, like one of Edge Hub or Edge Agent (not both, since we only want a single instance).

Mounting a config file is a bit tricky (need to deal with mount paths and making sure the Agent is well configured), I think since we're using the Docker Agent using autodiscovery is preferable. Another con to mounting a file is that device operators will most likely define everything through the Azure IoT Edge web UI, and there's no way to define arbitrary files there (not AFAIK at least). Defining labels in the Create Options should sound more familiar to device operators.

>                     "com.datadoghq.ad.check_names": "[\"azure_iot_edge\"]",

Author: Okay, moved these labels to the Edge Agent container :+1:

                    "com.datadoghq.ad.init_configs": "[{}]",
                    "com.datadoghq.ad.instances": "[{\"edge_hub_prometheus_url\": \"http://edgeHub:9600/metrics\", \"edge_agent_prometheus_url\": \"http://edgeAgent:9600/metrics\", \"security_manager_management_api_url\": \"unix:///var/run/iotedge/mgmt.sock\"}]"
                }
            }
            ```
        - Windows:
            ```json
            {
                "HostConfig": {
                    "NetworkMode": "default",
                    "Env": ["NetworkId=azure-iot-edge"],
                    "Binds": ["/var/run/docker.sock:/var/run/docker.sock"]
                },
                "Labels": {
                    "com.datadoghq.ad.check_names": "[\"azure_iot_edge\"]",
                    "com.datadoghq.ad.init_configs": "[{}]",

>                     "com.datadoghq.ad.instances": "[{\"edge_hub_prometheus_url\": \"http://edgeHub:9600/metrics\", \"edge_agent_prometheus_url\": \"http://edgeAgent:9600/metrics\", \"security_manager_management_api_url\": \"http://host.docker.internal:15580/\"]}]"

Author: 🤔 Need to think about it. Interestingly this worked on my Linux VM too, so…? But you're right, this is not documented anywhere for Docker on Linux (at least not that I could find).

I think there's a small discussion to be had related to this "check the security manager health by pinging on its management API running on the host" idea, though… And whether we need a service check for the security manager at all, at least in an initial release, since that's proving to cause so many configuration / compatibility problems. 🤔

>                     "com.datadoghq.ad.instances": "[{\"edge_hub_prometheus_url\": \"http://edgeHub:9600/metrics\", \"edge_agent_prometheus_url\": \"http://edgeAgent:9600/metrics\", \"security_manager_management_api_url\": \"http://host.docker.internal:15580/\"]}]"

Author: Dropped the security manager service check, as per offline discussion!
